src/Limits/Product.lidr

```diff
@@ -22,21 +22,87 @@ along with this program.  If not, see <https://www.gnu.org/licenses/>.
 > module Limits.Product
 >
 > import Basic.Category
+> import Basic.Functor
 > import Basic.Isomorphism
+> import Product.ProductCategory
 > import public CoLimits.CoProduct
 > import public Dual.DualCategory
->
+> 
 > %access public export
 > %default total
->
+> 
 > Product : (cat : Category) -> (a : obj cat) -> (b : obj cat) -> Type
 > Product cat a b = CoProduct (dualCategory cat) a b
->
+> 
 > productsAreIsomorphic :
 >      (a, b : Product cat l r)
 >   -> Isomorphic cat (carrier a) (carrier b)
 > productsAreIsomorphic {cat} {l} {r} a b =
 >   dualPreservesIsomorphic (coProductsAreIsomorphic (dualCategory cat) l r a b)
->
+> 
 > BinaryProduct : Category -> Type
 > BinaryProduct cat = BinaryCoProduct (dualCategory cat)
+> 
+> binaryProductToFunctor :
+>   BinaryProduct cat -> CFunctor (productCategory cat cat) cat
+> binaryProductToFunctor {cat} p = MkCFunctor
+>   mapObj
+>   (\a, b => mapMor {a} {b})
+>   preserveId
+>   (\a, b, c => preserveCompose {a} {b} {c})
+> where
+>   mapObj : obj (productCategory cat cat) -> obj cat
+>   mapObj (l, r) = carrier (p l r)
+>   
+>   mapMor : mor (productCategory cat cat) a b -> mor cat (mapObj a) (mapObj b)
+>   mapMor {a=(al, ar)} {b=(bl, br)} (MkProductMorphism pi1 pi2) =
+>     challenger
+>       (exists (p bl br)
+>               (carrier (p al ar))
+>               (compose cat (carrier (p al ar)) al bl (inl (p al ar)) pi1)
+>               (compose cat (carrier (p al ar)) ar br (inr (p al ar)) pi2))
+> 
+>   preserveId :
+>        (a : obj (productCategory cat cat))
+>     -> mapMor {a=a} {b=a} (identity (productCategory cat cat) a)
+>        = identity cat (mapObj a)
+>   preserveId (l, r) =
+>     rewrite unique (p l r)
+>               (carrier (p l r))
+>               (inl (p l r))
+>               (inr (p l r))
+>               (idCommutingMorphism (dualCategory cat) l r (p l r)) in
+>     rewrite rightIdentity cat (carrier (p l r)) l (inl (p l r)) in
+>     rewrite rightIdentity cat (carrier (p l r)) r (inr (p l r)) in Refl
+> 
+>   preserveCompose :
```

Comment:
I am a bit stuck on completing this definition. Any hints would be appreciated!

Reply:
@Jake-Gillberg it looks like you need to use something like

```
commutativityLeft (exists (p bl br)
                                        (carrier (p al br))
                                        (compose cat (carrier (p al ar)) al bl (inl (p al ar)) fpi1)
                                        (compose cat (carrier (p al ar)) ar br (inr (p al ar)) fpi2)
```

and the analogous `commutativityRight` for the other side